### PR TITLE
Connection timeout + check API integration

### LINKED ISSUE #46: Add timeout for requests
I have run into a situation where a restored config from another environment results in Moodle upgrade hanging at the search_elastic step as it tries and fails to reach the elastisearch server. Having a configurable timeout would resolve this issue where the elastisearch server falls over and/or cannot be reached.

### LINKED ISSUE #107: The status check should only occur on the settings page itself
It appears as though this check occurs on every single admin page. There should be a check around the status check call, so it only occurs on the plugin settings page

https://github.com/catalyst/moodle-search_elastic/blob/MOODLE_310_STABLE/settings.php#L31-L38

Its probably not an issue on a staging/production site, where this elasticsearch should be configured properly, but this killed site locally (after restore) till we were able to debug and pin-point the root cause.

E.g of such check in another plugin: https://github.com/ncstate-delta/moodle-mod_zoom/blob/main/settings.php#L38

Tasks:
* [ ] Turn this into a proper status check
* [ ] The check should have a very short curl timeout line 5 seconds
* [ ] Leverage the check in a micro ajax call
* [ ] Wrap this into an admin setting in this plugin
* [ ] Also prepare the same code to made generic and put into core:
https://tracker.moodle.org/browse/MDL-67898

### PR DESCRIPTION
**Pull request by matthewhilton**
## TLDR

- Adds connection timeout Closes #46
- Cleans up status check Closes #107

## Changes
- Configurable Guzzle http timeout set to default 5 seconds.
- Refactored the engine status check to provide better error messages (e.g. URL gave http code XYZ)
- Implemented check API https://moodledev.io/docs/apis/subsystems/check
- Created admin setting which takes in a check class and displays it nicely. (Note this will eventually be integrated a bit nicer likely with ajax in core https://tracker.moodle.org/browse/MDL-67898)
- This check now only runs when you are viewing the full tree (i.e. not searching) and when you are on the elasticsearch config page. Since it can take up to the timeout (5 seconds default) - we don't want this running when viewing other pages, etc...

## Screenshots

See https://github.com/catalyst/moodle-search_elastic/pull/108#discussion_r1325128349

### CODE REVIEW

File: classes/admin_setting_check.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+     * @param string $query
+     * @return string html
+     */
+    public function output_html($data, $query = '') {
+        global $OUTPUT;
+
+        // Run the check and get the result.
+        $checkresult = $this->check->get_result();
+
+        $resulthtml = $OUTPUT->check_result($checkresult);
```
Comment (brendanheywood) [inline classes/admin_setting_check.php:77]: Can you please make this a notification, success / warn / error etc

Comment (matthewhilton) [reply classes/admin_setting_check.php:79]: With 76b9243ca45219545437a089ce50fe47ca6c49e9 I have updated the formatting to use the notification. Note I only included error and ok notifications, since this check only uses n/a, error and ok check results.

In the core version it will support the rest e.g. warning, info, etc..

image
image
